### PR TITLE
Purchases: Make Refund Window Clearer 

### PR DESCRIPTION
**Pull request by Aurorum**
Related to #88979 (approved by Marketing there) 

## Proposed Changes

The intention here is to make it more clear when a user is eligible for a refund immediately from the Purchases screen. If the user clicks the link, they'll get some more detailed information about what this refund involves. However, many users don't get there (see the original issue), which has led to an increase in chargebacks and support requests related to this issue. 

## Testing Instructions

You can see this under Me > Purchases and selecting a purchase. There are three cases to test.

**1) When a user is eligible for a financial refund**
For example, when a user has just purchased a plan and it's still within the refund window. Something like this should appear: `Eligible for a refund of £4 until 12 April 2024.`

~~**2) When a user is eligible for a free domain credit**
For example, when a user has just registered a new domain with their plan within the last 96 hours, and would receive a new domain credit if they cancelled it. Something like this should appear: `If you cancel this domain before 9 April 2024, you will be able to choose another free domain.`~~

**3) When the user is ineligible for a refund**
For example, when over 14 days have passed since the purchase of an annual plan or 96 hours since a domain (or a non-refundable purchase in general). Nothing should show in this case; there is no difference to trunk. 

Screenshot 2024-04-05 at 12 14 44

cc @sirbrillig, @oswian, @jjchrisdiehl

### CODE REVIEW

**Comment by Aurorum** [comment]
I've updated the strings in-line with @alvezfrancese's suggestions here: https://github.com/Automattic/wp-calypso/issues/88979#issuecomment-2039897701

File: client/me/purchases/manage-purchase/index.tsx
Comment (jjchrisdiehl) [inline client/me/purchases/manage-purchase/index.tsx]: It is a good practice to avoid using `let` to define a variable since its definition can be redefined at any point later on by another developer. It also makes it more difficult to understand what the value of `refundText` is without reading through the full block (which may grow and change over time). @sirbrillig has a great explanation of this https://github.com/Automattic/wp-calypso/pull/88054#discussion_r1518413997.

TLDR you can use an IIFE to conditionally set `const refundText`:

```
const refundText = (() => {
  if ((value === 1)) {
    return "Refund text one!";
  }

  if ((value === 2)) {
    return "Refund text two!";
  }
  
  return "Default value";
})();
```

**Review by jjchrisdiehl — CHANGES_REQUESTED** [review]

File: client/me/purchases/manage-purchase/index.tsx
Comment (Aurorum) [reply client/me/purchases/manage-purchase/index.tsx]: Thanks for the review! Happy to make that change - I'd argue that this function is clearer than the one you linked, but I hear what you say about refund options increasing in future.  

Just a heads up that I don't have a paid subscription anymore to test this change, since I had to cancel mine so that I wouldn't be charged. I've tried testing by mocking some test data, but would be grateful if you could check that everything looks okay on Calypso!

File: client/me/purchases/manage-purchase/index.tsx
Comment (sirbrillig) [inline client/me/purchases/manage-purchase/index.tsx]: Super minor but we don't need the optional chaining here because there's a guard for `purchase` above so it won't ever be falsy here. It might be wise in fact to remove it because `new Date(undefined)` produces an invalid date object and if the guard above were ever removed, this could cause hidden bugs.

```suggestion
		const lastRenewalDate = new Date( purchase.mostRecentRenewDate );
```

File: client/me/purchases/manage-purchase/index.tsx
Comment (sirbrillig) [inline client/me/purchases/manage-purchase/index.tsx]: Similar to above, I don't think it's possible for `purchase` to be falsy here so we can remove the optional chaining.

```suggestion
			lastRenewalDate.getTime() + purchase.refundPeriodInDays * ONE_DAY_IN_MILLISECONDS
```

File: client/me/purchases/manage-purchase/index.tsx
Comment (sirbrillig) [inline client/me/purchases/manage-purchase/index.tsx]: Same here. We don't want to allow displaying `Elligible for a refund of undefined`.

```suggestion
						refundAmount: purchase.refundText,
```

File: client/me/purchases/manage-purchase/index.tsx
Comment (sirbrillig) [inline client/me/purchases/manage-purchase/index.tsx]: Same here. `refundPeriodInDays` is typed as a number but this allows it to be undefined, so that if the code above here changed in the future to allow a falsy `purchase` object through, this would happily print `The refund period for this purchase is undefined days`, which we can prevent by letting TypeScript guard against it:

```suggestion
									refundPeriodInDays: purchase.refundPeriodInDays,
```

**Review by sirbrillig — COMMENTED** [review]
Just some minor suggestions but I think they're worth doing. Then I'll put a string freeze on this to get the translations in.

File: client/me/purchases/manage-purchase/index.tsx
Comment (Aurorum) [reply client/me/purchases/manage-purchase/index.tsx]: Ah, thanks so much @sirbrillig! Those confused me, so I appreciate the extra explanation too. 

All fixed now. :)

**Comment by Aurorum** [comment]
Ah, thanks so much @sirbrillig! Those confused me, so I appreciate the extra explanation too. 

All fixed now. :)

File: client/me/purchases/manage-purchase/index.tsx
Comment (DavidRothstein) [inline client/me/purchases/manage-purchase/index.tsx]: I'm not sure this is an accurate way to figure out if a domain credit was used.  I think it could also pick up other cases where a domain was free (or even paid but then already refunded).

File: client/me/purchases/manage-purchase/index.tsx
Comment (DavidRothstein) [inline client/me/purchases/manage-purchase/index.tsx]: I don't believe this text is accurate.  Domain credits are one-time use; you can't use them over and over again.

File: client/me/purchases/manage-purchase/index.tsx
Comment (DavidRothstein) [inline client/me/purchases/manage-purchase/index.tsx]: Although I like the idea of mentioning the date, I think one reason we may not do that in the current UI (and rather use wording like "within X days of purchasing") is to avoid ambiguity over timezones, time of day, etc.  The situation we would want to avoid is to tell people that they're still eligible for a refund on a particular day, but then at some point before that day ends, the refund option has actually disappeared already.

**Review by DavidRothstein — CHANGES_REQUESTED** [review]
@Aurorum Thank you so much for working on these kinds of changes!

>approved by Marketing there

Can you clarify where someone approved this?  I see discussion there, but it wasn't clear to me where someone directly approved going forward with the change.

I think the change does have some potential benefits, but it also adds a fair amount of extra detail to this page that doesn't seem relevant for someone who has no intention of cancelling.  I'd personally prefer going back to having the link say something like "Cancel and refund plan" (i.e. undo the changes from https://github.com/Automattic/wp-calypso/pull/68088), which would communicate to them that they are currently eligible for a refund, but continue to require them to click through to get more details about the refund.  But that's just my opinion.

I also found some issues (and at least one actual inaccuracy) in the inline review comments.

File: client/me/purchases/manage-purchase/index.tsx
Comment (Aurorum) [reply client/me/purchases/manage-purchase/index.tsx]: Ah, that makes sense. Would you have any suggestions about the best way to check this? Although I suppose that wouldn’t be necessary if domain credits aren’t returned (see below!).

File: client/me/purchases/manage-purchase/index.tsx
Comment (Aurorum) [reply client/me/purchases/manage-purchase/index.tsx]: Sorry - silly question! - would you mind clarifying whether “believe” here means they’re definitely not refundable or would it be better to check with someone else? I haven’t tested since I don’t have a free domain credit on my account, but I asked my HE Advisor this week who said that “cancelling the domain should open up a new domain credit”.

File: client/me/purchases/manage-purchase/index.tsx
Comment (Aurorum) [reply client/me/purchases/manage-purchase/index.tsx]: My personal view is that the problem with mentioning the date is that people often still don’t realise what that means in practice (even when we try make it clear!). :) We could always include the time that the refund window expires here to alleviate that concern?

Happy to defer to you though.

**Comment by Aurorum** [comment]
Thanks for the review @DavidRothstein!

> Can you clarify where someone approved this? I see discussion there, but it wasn't clear to me where someone directly approved going forward with the change.

I was basing it of the original issue and the comment here: https://github.com/Automattic/wp-calypso/issues/88979#issuecomment-2036958162.

> I think the change does have some potential benefits, but it also adds a fair amount of extra detail to this page that doesn't seem relevant for someone who has no intention of cancelling

I might be biased here because I see this issue pop-up almost everyday in support tickets, but I’d feel that this detail is justified:

- I suspect that if you’ve ended up on the Managing Purchases page for a specific subscription, it’s probably likely that you have at least some intention of cancelling.
- It is temporary and it will disappear after the refund window has lapsed, so it’s not unnecessary detail!
- We often deal with tickets that are beyond the refund window. As a courtesy, if it’s only a few days over, these are sometimes refunded as a one-off by Support, but I suspect it would alleviate the Support backlog fairly significantly if the actual end refund date were made more prominent. 

As ever, happy to defer to others, but those are just my views. :)

File: client/me/purchases/manage-purchase/index.tsx
Comment (sirbrillig) [reply client/me/purchases/manage-purchase/index.tsx]: I'd have to search the refund code but I'm pretty sure David is right. If you've used a domain credit, I don't think there is code anywhere that restores it just because you've cancelled the domain within the refund period.

Comment (Aurorum) [reply client/me/purchases/manage-purchase/index.tsx]: Interesting! That seems like it would be useful, but I've removed that line from the code here for now.

Comment (DavidRothstein) [reply client/me/purchases/manage-purchase/index.tsx]: Yes, I tested it now and confirmed that the domain credit isn't restored after the domain is cancelled.  You could point your HE Advisor to this internal documentation which explains it in more detail: P6Ya1B-xS-p2

I thought there was public documentation about this too, but perhaps not... That may be because it's not necessarily a hard and fast rule (at least when HEs are involved), but in terms of what customers are allowed to do on their **own** in Calypso, it's intentional that they are restricted from using the domain credit over and over again.

File: client/me/purchases/manage-purchase/index.tsx
Comment (DavidRothstein) [reply client/me/purchases/manage-purchase/index.tsx]: I wonder if including the exact time might be **too** specific (even if it's an accurate countdown for when the option disappears from Calypso, we wouldn't necessarily always be so strict in practice).

Overall, I feel like it might just be best to stick to the "within X days of purchase or renewal" kind of wording (which is also the wording that was chosen for the documentation at https://wordpress.com/support/refunds/#refund-periods)?  People should be able to figure out when they purchased something -- and if it's not easy for them to figure that out, maybe **that's** the information we should actually be making easier for them to find :smile:

**Comment by DavidRothstein** [comment]
>I was basing it of the original issue and the comment here: https://github.com/Automattic/wp-calypso/issues/88979#issuecomment-2036958162.

Unfortunately I think that was more like a flyby editorial comment/text suggestion than anything else... Overall, the Marketing team itself has been very hesitant to promote refunds much at all in the past (hence things like https://github.com/Automattic/wp-calypso/pull/68088 which I mentioned earlier).  I would like to see things change, but feel like it might require more internal discussion to make sure people are in agreement.

I think I can get behind the idea of mentioning the refund end date/duration on this page, given that it's a bit of a "deadline" and I guess we usually don't like to make people click through to sub-screens to find those kinds of deadlines.  But I wouldn't lean towards adding much else; the main point of this screen isn't specifically for people to cancel their subscription, and we don't put lots of details underneath the other options on this page, but rather assume they'll click through to get it.

So for example, could it be as simple as this?
>Cancel plan
>...Eligible for a refund until [date]

Or (depending on the resolution of other questions above) more like this:
>Cancel plan
>...Eligible for a refund within 14 days of purchase

That seems to me to provide the most important info, and the rest isn't necessary here.  The refund amount should already be on the next page (if they click through) and adding a link to the documentation is a good idea, but again it could be on the next page after they click through.

The above is my opinion, but I also suspect that the less text added here, the more likely it is to get in.

**Comment by Aurorum** [comment]
Thanks for the review @DavidRothstein! 

> Overall, the Marketing team itself has been very hesitant to promote refunds much at all in the past 

Yeah, I did suspect that. It is wholly understandable, but it's sometimes a little unfortunate from a customer experience (and, based on the original issue, from an a8c experience in terms of dealing with chargebacks). 

> So for example, could it be as simple as this?

I think both your options are very good. Would you need someone to sign it off internally? It could also literally just be: `This purchase is still eligible for a refund.` and the tooltip could contain the information about the refund window length - that would keep it as tidy/minimal as possible. 

> I guess we usually don't like to make people click through to sub-screens to find those kinds of deadlines.

For what it's worth, I think the problem is actually getting people to that screen. From my experience, "Cancel plan" sounds scary to users and many won't click it without knowing that a refund is involved (hence why I think the text beneath it could make a lot of difference!). 

Thanks again. :)

File: client/me/purchases/manage-purchase/index.tsx
Comment (Aurorum) [reply client/me/purchases/manage-purchase/index.tsx]: I have no strong view on this (not that it would be a decision for me anyway!), but my general assumption is that if we can take a step out of the way for the user, that's even better. So it might be nice to make it easy for the user not to need to calculate the time which the refund period ends - also should lead to less aggrieved users in Support and refunds that way. 

But I also recognise more convenient might not always be better, especially in the case of refunds! 

Let me know if you think I should update it either way.

File: client/me/purchases/manage-purchase/index.tsx
Comment (Aurorum) [reply client/me/purchases/manage-purchase/index.tsx]: Yes, from my experience, the domain credit is always returned when people ask - but also makes sense that they might need to go through that process to prevent abuse etc. 

Thanks for checking!

**Comment by DavidRothstein** [comment]
>Would you need someone to sign it off internally?

Sorry for the delay here. I've posted an internal discussion about this (pebzTe-14a-p2) and one way or another hope we can have more of an answer in the next week or so.  Thank you for your patience!

As a side note, just wanted to mention that if we do wind up making changes to this UI, it may require some documentation updates as a followup (the previous ones that removed the "refund" mention in the first place were done here: https://github.com/Automattic/en.support-docs-content/issues/1705).

**Comment by DavidRothstein** [comment]
Hi @Aurorum, sorry for the extremely long delay here.

But, we are good to go ahead and implement this!  So if you (or anyone else) is still interested in working on it, please feel free.

One main refinement: Let's avoid the word "eligible".  (As you pointed out above, there are rare cases where someone might still be eligible for a refund even if they can't self-refund through the user interface, so it's probably better to avoid that word.)

I think we can therefore do either of these options:
1. "Refund available until [date]"
2. "Refund available within [X] days of purchase"

If going with the first option, as mentioned before we'd need to be careful not to overpromise what happens on the end date.  Obviously (as you noted) we aren't really going to be strict about it down to the minute (in the case where someone contacts a Happiness Engineer to ask for a refund), but it would still be disconcerting for a customer to think they have a self-serve refund option available and then have it disappear from the interface earlier than they expect it to, such that their only remaining recourse is to contact support for help.  My suggestion is as follows (others certainly welcome):
1. Convert the date at which it stops being refundable to the current timezone. Let's say that's on July 30.
2. If today is July 29 or earlier, the text can say "Refund available until July 30".
3. If today is July 30 (and the refund window hasn't expired yet), the text can say "Refund window ends today".

Hopefully that combination means they won't be surprised if and when the option disappears at some random time during the day on July 30.

Finally, regarding adding a link to the documentation, etc., my rough thought is that it's probably better to do that on the next page (after they click through to cancel) but I suppose if it's hidden behind a tooltip, doing it on this page isn't so bad as an alternative.

**Comment by Aurorum** [comment]
Thanks @DavidRothstein! Those options sound great and I'd be more than happy to implement. I'm afraid that I've run out of credits to test my code changes here though - would you mind adding either some credits or a plan subscription within the refund window to my WPcom account again? Sorry, and thanks again!!

**Comment by DavidRothstein** [comment]
Thanks so much, @Aurorum!  I can definitely give you credits or a plan subscription if you wind up needing it, but we actually have a new (and hopefully better) mechanism for testing purchases now.  I've sent some information about it to your WordPress.com account email address now.

**Comment by Aurorum** [comment]
Thanks @DavidRothstein! The sandbox is so cool, and definitely useful for making test sites - really appreciate you sharing it with me! The slight problem is that it makes it a little tricky to mock the expiry date for plans, but I think that this PR should still work nevertheless. 

I've made the changes which you referenced regarding the text. I don't really have strong views on the tooltip, so I'm happy to remove it - that said, I do feel like the cancellation page could do with a lot more love and it currently looks out-of-place (perhaps one for another PR if you'd be willing to review?). In the meantime, I've fixed a slight a11y colour contrast issue on that page too. 

Let me know how these changes seem now. :)

File: client/me/purchases/cancel-purchase/style.scss
Comment (DavidRothstein) [inline client/me/purchases/cancel-purchase/style.scss]: Is this the one you meant to change, or is it the one further up (under `.cancel-purchase__card`)?

It's possible I'm missing it, but I can't seem to find any difference as a result of this change.

File: client/me/purchases/manage-purchase/index.tsx
Comment (DavidRothstein) [inline client/me/purchases/manage-purchase/index.tsx:678]: Should the order of the checks be switched here, so that in the case where `purchase` is empty we don't try to pass it into `hasAmountAvailableToRefund()`?

You could do this either by just switching the order within the existing `if()` statement, or breaking it up into two separate `if()` statements each with its own early return.

File: client/me/purchases/manage-purchase/index.tsx
Comment (DavidRothstein) [inline client/me/purchases/manage-purchase/index.tsx]: This duplicates the same thing a few lines above.

File: client/me/purchases/manage-purchase/index.tsx
Comment (DavidRothstein) [inline client/me/purchases/manage-purchase/index.tsx]: This looks unused.

File: client/me/purchases/manage-purchase/index.tsx
Comment (DavidRothstein) [inline client/me/purchases/manage-purchase/index.tsx]: This looks unused.

File: client/me/purchases/manage-purchase/index.tsx
Comment (DavidRothstein) [inline client/me/purchases/manage-purchase/index.tsx]: Should there be an early return here if `refundText` is still empty?

In practice that shouldn't happen, since if the subscription isn't refundable it would have already returned early above.  But since `refundText` is being defined in an `if/elseif` then it's at least theoretically possible that it never gets set to anything, so it might be worth protecting against.

File: client/me/purchases/manage-purchase/index.tsx
Comment (DavidRothstein) [inline client/me/purchases/manage-purchase/index.tsx:731]: I think this one should be removed.  In practice it doesn't display anything (because the `RemovePurchase` component should only ever appear when the purchase isn't refundable in the first place), but if for some reason it does appear, we wouldn't want to tell people that a refund is available here, because going through the "Remove" flow (unlike the "Cancel" flow) will never trigger a refund.

File: client/me/purchases/manage-purchase/index.tsx
Comment (DavidRothstein) [inline client/me/purchases/manage-purchase/index.tsx]: When @michaeldcain was looking at this a while ago, he was wondering if it was OK to calculate this client-side, vs. having the server return it as part of the API.  I was hoping the answer was that it didn't matter, but after some investigation it appears there are cases where this calculation won't work correctly.

One example is plan downgrades.  I found that if you buy a Premium plan, let's say on August 1, and then downgrade to a Personal plan on August 5 (getting a partial refund in the process), then go to cancel the Personal plan and get the remainder of the charge refunded, the code here will tell you that you have until August 19 to get the refund.  But it's actually only until August 15 (14 days from when you made the payment).

Unfortunately, there's no real way to address that which I can think of, except to calculate the end date server-side... So although I'd hate to postpone this pull request any longer, I've created an issue for us to look into returning that data from the server (https://github.com/Automattic/payments-shilling/issues/3036).  We'll see what we can do.  If/when that happens, the code here would be able to access something like `purchase.refundPeriodEndDate` directly instead.

**Review by DavidRothstein — COMMENTED** [review]
@Aurorum, thanks again for your work on this!  I've reviewed it and left some inline comments.  Overall it looks good and seems to work well for the basic cases.  But one of the issues I found unfortunately might require server-side changes to fix.

**Comment by Aurorum** [comment]
Thanks for the review @DavidRothstein! Happy to make these changes, but might be best to wait for that API update - otherwise perhaps a more generic line like `Eligible for a refund within X days of purchase` might make sense? Curious to hear your thoughts.

Alsooo, if you could help me with either another sandbox cookie or some credits (I realised from #93211 that although the sandbox is great, it won't let me test purchases outside the refund window), that would be super appreciated and helpful for testing both those PRs! :)

**Comment by DavidRothstein** [comment]
>Alsooo, if you could help me with either another sandbox cookie or some credits

I gave you both, and sent the details to you by email.

>Happy to make these changes, but might be best to wait for that API update - otherwise perhaps a more generic line like `Eligible for a refund within X days of purchase` might make sense? Curious to hear your thoughts.

Yeah, at this point maybe worth waiting a little bit to see what happens with the API update first.  I know @jjchrisdiehl did start looking into the API update a little bit, so let's see how that goes and if we have time to finish it soon.

**Comment by jjchrisdiehl** [comment]
Hello @Aurorum!

Don't worry, I _haven't forgotten about you_. We've been going back and forth working on a reliable way to return the refund end date in our Purchases api, but have run into a number of scenarios that weren't easily accounted for with our current refund code. 

With that said, we're looking at a larger refactor of our refund code which will allow us to create a more robust system for returning this data.

At this point, we'll need to put this PR on the back burner until that work is completed. There isn't an estimated release date at this time, but we can reopen this PR once we have a proper fix available. Thanks for working on this and for your patience!

**Comment by DavidRothstein** [comment]
>At this point, we'll need to put this PR on the back burner until that work is completed.

That said, if we wanted to use language more like "Refund available" in the meantime (rather than trying to display the specific date) that is probably something we could consider moving forward with now.

Regarding the date, one issue is a technical one (actually being able to reliably and confidently return the correct date from the API) but the other is actually a conceptual one, and that one could be worth thinking about earlier... Specifically, what should we do when there is more than one date?  An example would be that the user purchases a plan along with a free domain.  In that case, it is possible for them to get a full refund of what they paid within the first 4 days, but between 4 and 14 days, they're only eligible for a partial refund, because the domain cost itself is no longer refundable, i.e. see https://wordpress.com/support/refunds/#domains ("If you register a free domain with your plan and request a refund after the 96-hour domain refund period, we will credit the difference between the cost of the plan and the cost of the domain").  So the question is, how would we communicate the refund window end date during those first 4 days?  One thought was that maybe its too complicated to communicate on this screen, so that simply displaying "Refund available" in that scenario is worth considering.

**Comment by Aurorum** [comment]
Thanks @jjchrisdiehl @DavidRothstein! No worries, that all makes sense - I've switched the text to just "Refund available" then, which should hopefully still help with the original reasoning of reducing confusion.

Screenshot 2024-09-28 at 13 28 06

> So the question is, how would we communicate the refund window end date during those first 4 days? 

That makes sense! I'd personally go with taking the latest date and then letting the user read the details on the next page - but I see how that can become complicated too. Hopefully this helps strike a good balance. 

Thanks again for all the help! :)

**Comment by DavidRothstein** [comment]
@Aurorum So sorry we missed your most recent work here!  We had it in a "Waiting for Something" queue (since it was originally waiting on the server-side changes) but missed that your most recent version isn't waiting on that anymore.  I'll put it in our review queue now and hopefully someone can take a look soon.

>With that said, we're looking at a larger refactor of our refund code

This refactor is actually complete -- we did it for several reasons, not just this issue, but it would make it easier to go ahead and return the refund window end date in the API if we need to, at least in some cases.  However, there's no need to wait for that here for now, since your pull request already represents an improvement as it is.

**Review by sirbrillig — APPROVED** [review]
Looks good!

**Comment by sirbrillig** [comment]
Adding a String Freeze for the translations in https://github.com/Automattic/wp-calypso/pull/97190. Once the translation is complete, we can merge this.

**Pull request merged**
